controller/ca/controller.go

```diff
@@ -0,0 +1,216 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type CertificateController struct {
+	namespace   string
+	k8sAPI      *k8s.API
+	syncHandler func(key string) error
+	queue       workqueue.RateLimitingInterface
+}
+
+func NewCertificateController(conduitNamespace string, k8sAPI *k8s.API) *CertificateController {
+	c := &CertificateController{
+		namespace: conduitNamespace,
+		k8sAPI:    k8sAPI,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(), "certificates"),
+	}
+
+	k8sAPI.Pod().Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.handlePodAdd,
+			UpdateFunc: c.handlePodUpdate,
+		},
+	)
+
+	k8sAPI.CM().Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.handleConfigMapAdd,
+			UpdateFunc: c.handleConfigMapUpdate,
+			DeleteFunc: c.handleConfigMapDelete,
+		},
+	)
+
+	c.syncHandler = c.syncNamespace
+
+	return c
+}
+
+func (c *CertificateController) Run(stopCh <-chan struct{}) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	log.Info("starting certificate controller")
+	defer log.Info("shutting down certificate controller")
+
+	go wait.Until(c.worker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *CertificateController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *CertificateController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err != nil {
+		log.Errorf("error syncing config map: %s", err)
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *CertificateController) syncNamespace(ns string) error {
+	conduitConfigMap, err := c.k8sAPI.CM().Lister().ConfigMaps(c.namespace).
+		Get(pkgK8s.CertificateBundleName)
+	if apierrors.IsNotFound(err) {
+		log.Warnf("configmap [%s] not found in namespace [%s]",
+			pkgK8s.CertificateBundleName, c.namespace)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	configMap := &v1.ConfigMap{
+		ObjectMeta: meta.ObjectMeta{Name: pkgK8s.CertificateBundleName},
+		Data:       conduitConfigMap.Data,
+	}
+
+	log.Debugf("adding configmap [%s] to namespace [%s]",
+		pkgK8s.CertificateBundleName, ns)
+	_, err = c.k8sAPI.Client.CoreV1().ConfigMaps(ns).Create(configMap)
+	if apierrors.IsAlreadyExists(err) {
+		_, err = c.k8sAPI.Client.CoreV1().ConfigMaps(ns).Update(configMap)
+	}
+
+	return err
+}
+
+func (c *CertificateController) handlePodAdd(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	if c.isInjectedPod(pod) && !c.filterNamespace(pod.Namespace) {
+		c.queue.Add(pod.Namespace)
+	}
+}
+
+func (c *CertificateController) handlePodUpdate(oldObj, newObj interface{}) {
+	c.handlePodAdd(newObj)
+}
+
+func (c *CertificateController) handleConfigMapAdd(obj interface{}) {
+	cm := obj.(*v1.ConfigMap)
```

Comment:
other places are doing something like:
```go
cm, ok := obj.(*v1.ConfigMap)
if !ok {
```
...any reason not to do that here?

Reply:
I can't seem to find written out explicitly anywhere, but the add and update funcs always receive an object, so it's safe to cast them without checking to make sure the cast succeeded. The delete func can receive a tombstone, in which case you have to have the check. From the cache docs:

```
* OnAdd is called when an object is added.
* OnUpdate is called when an object is modified. Note that oldObj is the
    last known state of the object-- it is possible that several changes
    were combined together, so you can't use this to see every single
    change. OnUpdate is also called when a re-list happens, and it will
    get called even if nothing changed. This is useful for periodically
    evaluating or syncing something.
* OnDelete will get the final state of the item if it is known, otherwise
    it will get an object of type DeletedFinalStateUnknown. This can
    happen if the watch is closed and misses the delete event and we don't
    notice the deletion until the subsequent re-list.
```

This convention also appears to be prevalent in the kubernetes code base. For instance, in the controller that I previously linked to, I see:

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/certificates/certificate_controller.go#L74